// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// The definition of validator RPC service.

syntax = "proto3";
package validator;
import "google/protobuf/struct.proto";

// Asset contains GCP resource metadata and additional metadata set on a resource, such as Cloud IAM policy.
message Asset {
  // GCP resource name as defined by Cloud Asset Inventory.
  // See https://cloud.google.com/resource-manager/docs/cloud-asset-inventory/resource-name-format for the format.
  string name = 1;
  // Cloud Asset Inventory type. Example: "google.cloud.sql.Instance" is the type of Cloud SQL instance.
  // This field has a redundant "asset" prefix to be consistent with Cloud Asset Inventory output.
  // See https://cloud.google.com/resource-manager/docs/cloud-asset-inventory/overview#supported_resource_types for the list of types.
  string asset_type = 2;
  // Ancestral project/folder/org information in a path-like format.
  // For example, a GCP project that is nested under a folder may have the following path:
  // organization/9999/folders/8888/projects/7777
  string ancestry_path = 3;
  // GCP resource metadata. The schema is type dependent.
  google.protobuf.Value resource = 4;
}

// Violation contains the relevant information to explain how a constraint is violated.
message Violation {
  // The name of the constraint that's violated.
  string constraint = 1;
  // GCP resource name. This is the same name in Asset.
  string resource = 2;
  // Human readable error message.
  string message = 3;
  // Metadata is optional. It contains the constraint-specific information that can potentially be used for remediation.
  // Example: In a firewall rule constraint violation, Metadata can contain the open port number.
  google.protobuf.Value metadata = 4;
}

message AddDataRequest {
  repeated Asset assets = 1;
}

message AddDataResponse {}

message AuditRequest {}

message AuditResponse {
  repeated Violation violations = 1;
}

message ResetRequest {}

message ResetResponse {}

service Validator {
  // AddData adds GCP resource metadata to be audited later.
  rpc AddData(AddDataRequest) returns (AddDataResponse) {}
  // Audit checks the GCP resource metadata that has been added via AddData to determine if any of the constraint is violated.
  rpc Audit(AuditRequest) returns (AuditResponse) {}
  // Reset clears previously added data from the underlying query evaluation engine.
  rpc Reset(ResetRequest) returns (ResetResponse) {}
}
